health_checker fixture fails with NoRunningCephToolBoxException after node disruption tests

```
2025-12-01 19:28:01request = <SubRequest 'health_checker' for <Function test_node_kernel_crash_ceph_fsync[CephFileSystem]>>
tier_marks_name = ['tier1', 'tier2', 'tier3', 'tier4', 'tier4a', 'tier4b', ...]
upgrade_marks_name = ['ocp_upgrade', 'mco_upgrade', 'dr_hub_upgrade', 'acm_upgrade', 'ocs_upgrade', 'pre_upgrade', ...]

    @pytest.fixture(scope="function", autouse=True)
    def health_checker(request, tier_marks_name, upgrade_marks_name):
        skipped = False
        ceph_health_recovered_exception = None
        dev_mode = ocsci_config.RUN["cli_params"].get("dev_mode")
        mcg_only_deployment = ocsci_config.ENV_DATA["mcg_only_deployment"]
        if mcg_only_deployment:
            log.info("Skipping health checks for MCG only mode")
            return
        if dev_mode:
            log.info("Skipping health checks for development mode")
            return
    
        if ocsci_config.multicluster:
            if (
                ocsci_config.cluster_ctx.MULTICLUSTER["multicluster_index"]
                == ocsci_config.get_active_acm_index()
            ):
                return
    
        node = request.node
    
        # Skip health check if the test is marked as 'Resiliency' or 'Chaos'
        skip_markers = ["resiliency", "chaos"]
        for mark in node.iter_markers():
            if mark.name.lower() in skip_markers:
                log.info(
                    f"Skipping Ceph health check for test marked with '{mark.name.title()}'"
                )
                return
    
        # ignore ceph health check for the TestFailurePropagator test cases
        if "FailurePropagator" in str(node.cls):
            return
    
        def finalizer():
            if not skipped:
                multi_storagecluster_external_health_passed = False
                try:
                    teardown = ocsci_config.RUN["cli_params"]["teardown"]
                    skip_ocs_deployment = ocsci_config.ENV_DATA["skip_ocs_deployment"]
                    ceph_cluster_installed = ocsci_config.RUN.get("cephcluster")
                    if not (
                        teardown
                        or skip_ocs_deployment
                        or mcg_only_deployment
                        or not ceph_cluster_installed
                    ):
                        # We are allowing 20 re-tries for health check, to avoid teardown failures for cases like:
                        # "flip-flopping ceph health OK and warn because of:
                        # HEALTH_WARN Reduced data availability: 2 pgs peering
                        ceph_health_check(
                            namespace=ocsci_config.ENV_DATA["cluster_namespace"],
                            fix_ceph_health=True,
                        )
                        log.info("Ceph health check passed at teardown!")
                        if ocsci_config.DEPLOYMENT.get("multi_storagecluster"):
                            ceph_health_check_multi_storagecluster_external()
                            log.info(
                                "Ceph health check for multi-storagecluster external cluster passed at teardown!"
                            )
                            multi_storagecluster_external_health_passed = True
    
                except CephHealthException:
                    if not ocsci_config.RUN["skip_reason_test_found"]:
                        squad_name = None
                        for marker in node.iter_markers():
                            if "_squad" in marker.name:
                                squad_name = marker.name
                                break
                        ocsci_config.RUN["skip_reason_test_found"] = {
                            "test_name": node.name,
                            "squad": squad_name,
                        }
                    log.info("Ceph health check failed at teardown")
                    # Retrying to increase the chance the cluster health will be OK
                    # for next test
                    ceph_health_check(namespace=ocsci_config.ENV_DATA["cluster_namespace"])
    
                    if (
                        not multi_storagecluster_external_health_passed
                        and ocsci_config.DEPLOYMENT.get("multi_storagecluster")
                    ):
                        ceph_health_check_multi_storagecluster_external()
                    raise
            if ceph_health_recovered_exception:
                """
                Handling a rare scenario where the teardown of the previous test completed successfully,
                but the setup of the next test detects an unhealthy Ceph state:
                - If a test case encounters a Ceph health issue during setup and it later recovers,
                  we still need to fail the test during teardown to make the issue noticeable.
                - However, if Ceph health recovered before running the test, we proceed with execution
                  to avoid losing test coverage.
                """
                log.error(
                    "We are failing the test case because it was not healthy in the setup phase"
                    " and the health status was forcibly recovered!"
                )
                raise ceph_health_recovered_exception
    
        request.addfinalizer(finalizer)
        for mark in node.iter_markers():
            if mark.name in tier_marks_name + upgrade_marks_name and ocsci_config.RUN.get(
                "cephcluster"
            ):
                log.info("Checking for Ceph Health OK ")
                external_multi_storagecluster_status = False
                try:
>                   status = ceph_health_check(
                        namespace=ocsci_config.ENV_DATA["cluster_namespace"],
                        tries=10,
                        delay=15,
                        fix_ceph_health=True,
                    )

tests/conftest.py:2111: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
ocs_ci/utility/utils.py:2754: in ceph_health_check
    return retry(
ocs_ci/utility/retry.py:99: in f_retry
    return f(*args, **kwargs)
ocs_ci/utility/utils.py:2787: in ceph_health_check_base
    health = run_ceph_health_cmd(namespace)
ocs_ci/utility/utils.py:2837: in run_ceph_health_cmd
    ct_pod = get_ceph_tools_pod(namespace=namespace)
ocs_ci/ocs/resources/pod.py:940: in get_ceph_tools_pod
    running_ct_pods = _get_tools_pod_objs()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    def _get_tools_pod_objs():
        """
        Method to get the Running ceph tool box pod objects
    
        Returns:
            List: of pod objects
    
        """
        with config.RunWithProviderConfigContextIfAvailable():
            ocp_pod_obj = OCP(
                kind=constants.POD,
                namespace=namespace,
                selector=constants.TOOL_APP_LABEL,
                cluster_kubeconfig=cluster_kubeconfig,
            )
            ct_pod_items = ocp_pod_obj.data["items"]
            logger.info(
                f"These are the ceph tool box pods: {[pod.get('metadata').get('name') for pod in ct_pod_items]}"
            )
            if not (ct_pod_items or skip_creating_pod):
                # setup ceph_toolbox pod if the cluster has been setup by some other CI
                setup_ceph_toolbox()
                ocp_pod_obj = OCP(
                    kind=constants.POD,
                    namespace=namespace,
                    selector=constants.TOOL_APP_LABEL,
                )
                ct_pod_items = ocp_pod_obj.data["items"]
    
            if not ct_pod_items:
                raise CephToolBoxNotFoundException
    
            if not get_running_pods:
                # Return the ceph tool pod objects even if they are not running
                return ct_pod_items
    
            # In the case of node failure, the CT pod will be recreated with the old
            # one in status Terminated. Therefore, need to filter out the Terminated pod
    
            # Update the OCP pod object without the selector
            ocp_pod_obj = OCP(
                kind=constants.POD,
                namespace=namespace,
                cluster_kubeconfig=cluster_kubeconfig,
            )
            running_ct_pods = list()
    
            for pod in ct_pod_items:
                pod_status = ocp_pod_obj.get_resource_status(
                    pod.get("metadata").get("name")
                )
                logger.info(f"Pod name: {pod.get('metadata').get('name')}")
                logger.info(f"Pod status: {pod_status}")
                if pod_status == constants.STATUS_RUNNING:
                    running_ct_pods.append(pod)
    
            if not running_ct_pods:
>               raise NoRunningCephToolBoxException
E               ocs_ci.ocs.exceptions.NoRunningCephToolBoxException

ocs_ci/ocs/resources/pod.py:931: NoRunningCephToolBoxException
2025-12-01 19:28:01def finalizer():
        if not skipped:
            multi_storagecluster_external_health_passed = False
            try:
                teardown = ocsci_config.RUN["cli_params"]["teardown"]
                skip_ocs_deployment = ocsci_config.ENV_DATA["skip_ocs_deployment"]
                ceph_cluster_installed = ocsci_config.RUN.get("cephcluster")
                if not (
                    teardown
                    or skip_ocs_deployment
                    or mcg_only_deployment
                    or not ceph_cluster_installed
                ):
                    # We are allowing 20 re-tries for health check, to avoid teardown failures for cases like:
                    # "flip-flopping ceph health OK and warn because of:
                    # HEALTH_WARN Reduced data availability: 2 pgs peering
>                   ceph_health_check(
                        namespace=ocsci_config.ENV_DATA["cluster_namespace"],
                        fix_ceph_health=True,
                    )

tests/conftest.py:2054: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
ocs_ci/utility/utils.py:2754: in ceph_health_check
    return retry(
ocs_ci/utility/retry.py:99: in f_retry
    return f(*args, **kwargs)
ocs_ci/utility/utils.py:2787: in ceph_health_check_base
    health = run_ceph_health_cmd(namespace)
ocs_ci/utility/utils.py:2837: in run_ceph_health_cmd
    ct_pod = get_ceph_tools_pod(namespace=namespace)
ocs_ci/ocs/resources/pod.py:940: in get_ceph_tools_pod
    running_ct_pods = _get_tools_pod_objs()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    def _get_tools_pod_objs():
        """
        Method to get the Running ceph tool box pod objects
    
        Returns:
            List: of pod objects
    
        """
        with config.RunWithProviderConfigContextIfAvailable():
            ocp_pod_obj = OCP(
                kind=constants.POD,
                namespace=namespace,
                selector=constants.TOOL_APP_LABEL,
                cluster_kubeconfig=cluster_kubeconfig,
            )
            ct_pod_items = ocp_pod_obj.data["items"]
            logger.info(
                f"These are the ceph tool box pods: {[pod.get('metadata').get('name') for pod in ct_pod_items]}"
            )
            if not (ct_pod_items or skip_creating_pod):
                # setup ceph_toolbox pod if the cluster has been setup by some other CI
                setup_ceph_toolbox()
                ocp_pod_obj = OCP(
                    kind=constants.POD,
                    namespace=namespace,
                    selector=constants.TOOL_APP_LABEL,
                )
                ct_pod_items = ocp_pod_obj.data["items"]
    
            if not ct_pod_items:
                raise CephToolBoxNotFoundException
    
            if not get_running_pods:
                # Return the ceph tool pod objects even if they are not running
                return ct_pod_items
    
            # In the case of node failure, the CT pod will be recreated with the old
            # one in status Terminated. Therefore, need to filter out the Terminated pod
    
            # Update the OCP pod object without the selector
            ocp_pod_obj = OCP(
                kind=constants.POD,
                namespace=namespace,
                cluster_kubeconfig=cluster_kubeconfig,
            )
            running_ct_pods = list()
    
            for pod in ct_pod_items:
                pod_status = ocp_pod_obj.get_resource_status(
                    pod.get("metadata").get("name")
                )
                logger.info(f"Pod name: {pod.get('metadata').get('name')}")
                logger.info(f"Pod status: {pod_status}")
                if pod_status == constants.STATUS_RUNNING:
                    running_ct_pods.append(pod)
    
            if not running_ct_pods:
>               raise NoRunningCephToolBoxException
E               ocs_ci.ocs.exceptions.NoRunningCephToolBoxException

ocs_ci/ocs/resources/pod.py:931: NoRunningCephToolBoxException
```


Teardown Flow Summary

  Test A Teardown:
    1. test's teardown fixture → nodes.restart_nodes_by_stop_and_start_teardown()
       └── Starts VMs/nodes BUT does NOT wait for:
           - Node Ready status
           - Rook operator pod Running
           - Toolbox pod recovery

    2. health_checker finalizer → ceph_health_check()
       └── May fail or succeed (timing dependent)

  Test B Setup:
    1. health_checker setup → ceph_health_check()
       └── get_ceph_tools_pod() → NoRunningCephToolBoxException

--- a/.secrets.baseline
+++ b/.secrets.baseline
@@ -590,6 +590,14 @@
         "line_number": 2810,
         "type": "Secret Keyword",
         "verified_result": null
+      },
+      {
+        "hashed_secret": "66a36e77fd002579809717841f998f4d21cd5913",
+        "is_secret": false,
+        "is_verified": false,
+        "line_number": 2811,
+        "type": "Secret Keyword",
+        "verified_result": null
       }
     ],
     "ocs_ci/krkn_chaos/template/krkn_global_config.yaml.j2": [

--- a/ocs_ci/helpers/helpers.py
+++ b/ocs_ci/helpers/helpers.py
@@ -43,6 +43,7 @@ from ocs_ci.ocs.utils import (
 from ocs_ci.ocs import constants, defaults, node, ocp, exceptions
 from ocs_ci.ocs.exceptions import (
     CommandFailed,
+    NoRunningCephToolBoxException,
     ResourceNotFoundError,
     ResourceWrongStatusException,
     TimeoutExpiredError,
@@ -57,6 +58,7 @@ from ocs_ci.utility.vsphere import VSPHERE
 from ocs_ci.utility.operators import NMStateOperator
 from ocs_ci.utility.retry import retry
 from ocs_ci.utility.utils import (
+    ceph_health_check,
     TimeoutSampler,
     ocsci_log_path,
     run_cmd,
@@ -2874,6 +2876,52 @@ def wait_for_ct_pod_recovery():
         else:
             return False
     return True
+
+
+def ceph_health_check_with_toolbox_recovery(
+    namespace: str,
+    tries: int = 20,
+    delay: int = 30,
+    fix_ceph_health: bool = True,
+) -> bool:
+    """
+    Perform ceph health check with automatic toolbox pod recovery.
+
+    If the ceph toolbox pod is not running (e.g., after node disruption tests),
+    this function will wait for the toolbox pod to recover before retrying.
+
+    Args:
+        namespace (str): Kubernetes namespace for ceph cluster.
+        tries (int): Number of retries for health check.
+        delay (int): Delay between retries in seconds.
+        fix_ceph_health (bool): Whether to attempt fixing ceph health issues.
+
+    Returns:
+        bool: True if ceph health check passes.
+
+    Raises:
+        NoRunningCephToolBoxException: If toolbox pod doesn't recover.
+        CephHealthException: If ceph health check fails after retries.
+    """
+    try:
+        return ceph_health_check(
+            namespace=namespace,
+            tries=tries,
+            delay=delay,
+            fix_ceph_health=fix_ceph_health,
+        )
+    except NoRunningCephToolBoxException:
+        logger.warning(
+            "Ceph toolbox pod not running. Waiting for recovery before retry..."
+        )
+        if wait_for_ct_pod_recovery():
+            return ceph_health_check(
+                namespace=namespace,
+                tries=tries,
+                delay=delay,
+                fix_ceph_health=fix_ceph_health,
+            )
+        raise
 
 
 def label_worker_node(node_list, label_key, label_value):

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -200,6 +200,7 @@ from ocs_ci.utility.utils import (
 from ocs_ci.helpers import helpers, dr_helpers
 from ocs_ci.helpers.helpers import (
     add_scc_policy,
+    ceph_health_check_with_toolbox_recovery,
     create_unique_resource_name,
     create_ocs_object_from_kind_and_name,
     setup_pod_directories,
@@ -2055,7 +2056,7 @@ def health_checker(request, tier_marks_name, upgrade_marks_name):
                     # We are allowing 20 re-tries for health check, to avoid teardown failures for cases like:
                     # "flip-flopping ceph health OK and warn because of:
                     # HEALTH_WARN Reduced data availability: 2 pgs peering
-                    ceph_health_check(
+                    ceph_health_check_with_toolbox_recovery(
                         namespace=ocsci_config.ENV_DATA["cluster_namespace"],
                         fix_ceph_health=True,
                         update_jira=True,
@@ -2083,7 +2084,9 @@ def health_checker(request, tier_marks_name, upgrade_marks_name):
                 log.info("Ceph health check failed at teardown")
                 # Retrying to increase the chance the cluster health will be OK
                 # for next test
-                ceph_health_check(namespace=ocsci_config.ENV_DATA["cluster_namespace"])
+                ceph_health_check_with_toolbox_recovery(
+                    namespace=ocsci_config.ENV_DATA["cluster_namespace"]
+                )
 
                 if (
                     not multi_storagecluster_external_health_passed
@@ -2114,7 +2117,7 @@ def health_checker(request, tier_marks_name, upgrade_marks_name):
             log.info("Checking for Ceph Health OK ")
             external_multi_storagecluster_status = False
             try:
-                status = ceph_health_check(
+                status = ceph_health_check_with_toolbox_recovery(
                     namespace=ocsci_config.ENV_DATA["cluster_namespace"],
                     tries=10,
                     delay=15,

